src/service_objects/game_piece_action_resolvers/melee_attack_resolver.ts

```diff
@@ -1,45 +1,99 @@
+  }
+
+  // Fetch some additional info
+  const attackingPlayerUnit = await Models.PlayerUnit.findByPk(attackingGamePiece.playerUnitId, { transaction: transaction });
+  const attackingUnit = await Models.Unit.findByPk(attackingPlayerUnit.unitId, { transaction: transaction });
 
-  if (commitChanges) {
-    // Validations done, modify state
-    if (targetGamePiece.isDead()) return attackingGamePiece;
+  return {
+    targetGamePiece,
+    attackingPlayerUnit,
+    attackingUnit,
+    distanceToTarget
+  };
+}
+
+export default async function resolveMeleeAttackAction(
+  action: Models.GamePieceAction,
+  transaction
+): Promise<Models.GamePiece> {
+  const actionData = action.actionData;
+  if (actionData.actionType !== 'meleeAttack') throw new Error(`Unable to resolve melee attack action with actionType: ${actionData.actionType}`);
+
+  const attackingGamePiece = await action.gamePiece();
+  const targetGamePieceId = actionData.targetGamePieceId;
+
+  // Validate action inputs and capture queried data
+  const {
+    targetGamePiece,
+    attackingUnit,
+  } = await validateMeleeAttackAction(
+    attackingGamePiece,
+    targetGamePieceId,
+    true,
+    transaction
+  );
+
+  // If target is already dead just abort
+  if (targetGamePiece.isDead()) return;
   
-    const attackingPlayerUnit = await Models.PlayerUnit.findByPk(attackingGamePiece.playerUnitId, { transaction: transaction });
-    const attackingUnit = await Models.Unit.findByPk(attackingPlayerUnit.unitId, { transaction: transaction });
-    const targetPlayerUnit = await Models.PlayerUnit.findByPk(targetGamePiece.playerUnitId, { transaction: transaction });
-    const targetUnit = await Models.Unit.findByPk(targetPlayerUnit.unitId, { transaction: transaction });
-
-    // TODO: Removed attackPower and armor, stub this as 1 for now.
-    //   Need to implement dice rolls and attack sequence.
-    // const damageSubtotal = attackingUnit.attackPower - targetUnit.armor;
-    const damageSubtotal = 1;
-    
-    const damage = Math.max(damageSubtotal, 1);
-    const newHealth = Math.max(targetGamePiece.health - damage, 0);
+  const attackingPlayerUnit = await Models.PlayerUnit.findByPk(attackingGamePiece.playerUnitId, { transaction: transaction });
+  const targetPlayerUnit = await Models.PlayerUnit.findByPk(targetGamePiece.playerUnitId, { transaction: transaction });
+  const targetUnit = await Models.Unit.findByPk(targetPlayerUnit.unitId, { transaction: transaction });
+
+  // Resolve attack sequence
+  const encodedDiceRolls = actionData.encodedDiceRolls;
+  const resolvedAttacks = resolveAttacks(
+    attackingUnit.meleeNumAttacks,
+    attackingUnit.meleeHitRoll,
+    attackingUnit.meleeWoundRoll,
+    targetUnit.armorSaveRoll,
+    attackingUnit.meleeArmorPiercing,
+    attackingUnit.meleeDamage,
+    encodedDiceRolls
+  );
+  const totalDamage = resolvedAttacks.reduce((sum, attack) => sum + attack.damageDealt, 0);
```

Comment:
Aha, I guess implicit in the decision to process the set of 3 dice rolls as individual sets is that there are now "actions" where the user says to attack, and "actions" where the piece takes a single swipe, but have more this turn.  Right now, how many actions a single piece can take in a single phase is actually not constrained in the proof, so this shouldn't be a problem until we patch that gap.

Reply:
Yeah apologies for overloading the term Action a bit.

A GamePieceAction in this case is a generic thing being done by a GamePiece, in this example one GamePiece attacking another. The fact that the unit likely has >1 "number of attacks" which results in multiple sets of dice being rolled I would say you don't have to treat as multiple distinct actions.

If my Swordsman attacks another unit and makes "three attacks" we aren't actually updating the game state after each attack, we're just resolving all the dice rolls, summing up the damage done, and applying the damage in one go. This simplifies things and is fine because there is not actually anything else that can happen in between each attack from the same attacking unit.

Would that distinction make the proof easier? Since you don't have to include a condition to see if the piece is dead in between each attack from the same attacking unit?